main/src/main/scala/sbt/internal/CommandExchange.scala

```diff
@@ -74,7 +74,7 @@ private[sbt] final class CommandExchange {
       def onIncomingSocket(socket: Socket): Unit =
         {
           s.log.info(s"new client connected from: ${socket.getPort}")
-          val channel = new NetworkChannel(newChannelName, socket)
+          val channel = new NetworkChannel(newChannelName, socket, s)
```

Comment:
NetworkChannel represents a connection from some user. I don't think a state object should be owned by it since it's sort of misleading since the "state" there represents a past state.
I think a better approach is to pass in `(onQuery: Query => Unit)` here that it can call back, and handle query in the `CommandExchange`.